### PR TITLE
Add servo calibration!

### PR DESCRIPTION
**Pull request by theicfire**

### CODE REVIEW

File: embedded/servo_driver/servo_driver.cc [hunk 1]
```diff
@@ -3,26 +3,52 @@
...
 #include <cmath> // floor
 #include <string.h> // memset
+#include <sstream>
+#include <iostream>
+#include <fstream>
+#include <yaml-cpp/yaml.h>
 
-ServoDriver::ServoDriver(char* dev) : pwm_driver_(dev) {}
 namespace {
   constexpr int PWM_FREQUENCY = 333;
```
Comment (theicfire) [inline embedded/servo_driver/servo_driver.cc]: remove

File: embedded/servo_driver/servo_driver.hh [hunk 2]
```diff
@@ -6,17 +6,28 @@
...
+	void set_calibrated_angle(float angle);
 
 private:
-        PwmDriver pwm_driver_;
+	int channel_;
+        PwmDriver &pwm_driver_;
+        std::string config_path_;
+	// The percentage corresponding to max angle.
+	int calibrated_max_;
+	// The percentage corresponding to zero angle.
```
Comment (theicfire) [inline embedded/servo_driver/servo_driver.hh]: center angle

File: embedded/servo_driver/servo_test.cc [hunk 3]
```diff
@@ -1,15 +1,42 @@
...
-        std::cout << "sweeping through servo" << std::endl;
-        for (int i = 0; i <= 100; ++i) {
-          driver.set_percentage(channel, i);
-          usleep(10000);
+	PwmDriver pwm_driver = PwmDriver("/dev/i2c-1");
+	pwm_driver.init();
+	pwm_driver.set_pwm_freq(330);
+	pwm_driver.enable_auto_increment(true);
+
+	ServoDriver servo1 = ServoDriver(1, pwm_driver, "cfg/servo_cfg1.yaml");
```
Comment (theicfire) [inline embedded/servo_driver/servo_test.cc]: 0, 1...

Comment (theicfire) [reply embedded/servo_driver/servo_test.cc]: change constructor

File: embedded/servo_driver/servo_test.cc [hunk 4]
```diff
@@ -1,15 +1,42 @@
...
-	ServoDriver driver = ServoDriver("/dev/i2c-1");
-	driver.init();
-        const uint8_t channel = 1;
-        std::cout << "sweeping through servo" << std::endl;
-        for (int i = 0; i <= 100; ++i) {
-          driver.set_percentage(channel, i);
-          usleep(10000);
+	PwmDriver pwm_driver = PwmDriver("/dev/i2c-1");
+	pwm_driver.init();
+	pwm_driver.set_pwm_freq(330);
```
Comment (theicfire) [inline embedded/servo_driver/servo_test.cc]: variable

Comment (theicfire) [reply embedded/servo_driver/servo_test.cc]: also in other file too

File: infrastructure/scripts/jet_run.sh [hunk 5]
```diff
@@ -6,16 +6,16 @@ if [ $? -ne 0 ]; then
     exit $?
 fi
 
-xhost +
+#xhost +
```
Comment (theicfire) [inline infrastructure/scripts/jet_run.sh]: undo these changes

Comment (theicfire) [reply infrastructure/scripts/jet_run.sh]: done

File: embedded/servo_driver/servo_test.cc (hunk 3 above)
Comment (theicfire) [reply embedded/servo_driver/servo_test.cc]: done

File: embedded/servo_driver/servo_driver.cc (hunk 1 above)
Comment (theicfire) [reply embedded/servo_driver/servo_driver.cc]: done

File: infrastructure/scripts/jet_run.sh [hunk 6]
```diff
@@ -8,7 +8,7 @@ fi
 
 xhost +
 
-CONTAINER_ID=$(docker run -it -d -v $JET_REPO_PATH:/jet -v /tmp/.X11-unix:/tmp/.X11-unix -e DISPLAY -e NO_AT_BRIDGE=1 --privileged hoverjet/jet bash)
+CONTAINER_ID=$(docker run -it -d -v $JET_REPO_PATH:/jet -v /tmp/.X11-unix:/tmp/.X11-unix -e DISPLAY -e NO_AT_BRIDGE=1 --privileged hoverjet/jet-arm bash)
```
Comment (benkroop) [inline infrastructure/scripts/jet_run.sh]: Undo change here too. I have  PR up that makes this switch automatic.

File: embedded/servo_driver/servo_test.cc [hunk 7]
```diff
@@ -1,15 +1,44 @@
 #include "servo_driver.hh"
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_test.cc]: Use the full path `embedded/servo_driver/servo_test.cc` so it can be found by pymake

File: embedded/servo_driver/servo_driver.cc [hunk 8]
```diff
@@ -3,26 +3,50 @@
...
+  try {
+    YAML::Node config = YAML::LoadFile(config_path_);
+    max_angle_ = config["max_angle"].as<int>();
+    calibrated_center_ = config["calibrated_center"].as<int>();
+    calibrated_max_ = config["calibrated_max"].as<int>();
+  } catch (YAML::BadFile e) {
+    max_angle_ = 20;
+    calibrated_center_ = 50;
+    calibrated_max_ = 100;
+  }
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.cc]: Suggest just crashing if an exception is thrown, instead of some unplanned behavior

File: embedded/servo_driver/servo_driver.hh [hunk 9]
```diff
@@ -6,17 +6,28 @@
...
+	int calibrated_center_;
+	int max_angle_;
+	int percentage_;
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.hh]: This is all uninitialized memory until `init()` is called; consider:

- Adding an `initialized_` bool, and doing `assert(initialized_)` in any function that uses these
- Or initializing them to something reasonable
- Or always calling `init()` in the constructor, so these are always initialized to a sane value

**Comment by theicfire** [comment]
Fixed building issues!

**Comment by theicfire** [comment]
@jaz-jlh @benkroop resolved all comments!

File: embedded/pwm_driver/pwm_driver.cc [hunk 10]
```diff
@@ -37,8 +37,9 @@ Adafruit PCA9685 library: https://github.com/adafruit/Adafruit-PWM-Servo-Driver-
 #define ALLLED_OFF_H 0xFD
 
 
-PwmDriver::PwmDriver(char* dev) {
-	device_name = dev;
+PwmDriver::PwmDriver(const std::string &dev) {
+	device_name = dev.c_str();
```
Comment (jpanikulam) [inline embedded/pwm_driver/pwm_driver.cc]: This may not do what you want; since this will get a pointer to a stack object that will be deallocated; I believe this is a dangling reference

You want to do one of the following:
- At the point where device_name_ is normally used, get the c_str
- Allocate a copy and hold onto that in the class

Comment (theicfire) [reply embedded/pwm_driver/pwm_driver.cc]: did neither, combined `init`

File: embedded/servo_driver/servo_driver.hh [hunk 11]
```diff
@@ -6,17 +6,28 @@
...
 	int init();
-	void set_percentage(int channel, int percentage);
+	void set_percentage(int percentage);
+	int get_percentage() const;
+	void set_angle(float angle);
 
 private:
-        PwmDriver pwm_driver_;
+	int channel_;
+        PwmDriver &pwm_driver_;
```
[the remark this answers is not in the document]
Comment (theicfire) [reply embedded/servo_driver/servo_driver.hh]: From talking over Slack, the way I'm using this is not an issue, but it could be used in a bad way.. it's important to ensure that pwm_driver_ *always* outlasts the lifetime of servo_driver

File: embedded/servo_driver/servo_driver.hh [hunk 12]
```diff
@@ -7,23 +7,24 @@
...
 	int init();
 	void set_percentage(int percentage);
 	int get_percentage() const;
 	void set_angle(float angle);
 
 private:
 	int channel_;
-        PwmDriver &pwm_driver_;
-        std::string config_path_;
+	const std::shared_ptr<PwmDriver> &pwm_driver_;
```
Comment (jpanikulam) [inline embedded/servo_driver/servo_driver.hh]: This reference is still dangling, since it's only a reference to the shared pointer, it wasn't copied, so the reference counter didn't get incremented

you likely want:
`const std::shared_ptr<PwmDriver> pwm_driver_` (no ampersand)

File: embedded/pwm_driver/pwm_driver.cc [hunk 13]
```diff
@@ -12,6 +12,7 @@ Adafruit PCA9685 library: https://github.com/adafruit/Adafruit-PWM-Servo-Driver-
 #include <algorithm> // min
 #include <cmath> // floor
 #include <string.h> // memset
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (theicfire) [reply embedded/pwm_driver/pwm_driver.cc]: used for `runtime_exception`
